File: dart/problem002.dart
import 'dart:async';
import 'dart:math';

import 'euler_solution.dart';

main(List<String> args) {
  EulerSolution.execute(new Problem001());
}

class Problem001 extends EulerSolution {

  @override
  Future<int> run() async {
    // ~200ms
    // Dynamically generate fibonacci sequence
    // it has an arbitrarily large max, we don't generate that much
    // var list = new Iterable<int>.generate(pow(2, 20), (i) => fib(i));

    // 20ms
    var list = fibStream();

    return list
      .takeWhile((i) => i <= 4000000)
      .where((i) => i.isEven)
      .reduce((a, b) => a + b);
  }

  int fib(int n) {
    if (n == 0 || n == 1)
      return n;
    else
      return fib(n - 1) + fib(n - 2);
  }

  Stream<int> fibStream() async* {
    int a = 0;
    int b = 1;

    while (true) {
      int curr = a + b;
      a = b;
      b = curr;
      yield curr;
    }
  }
}

File: c/problem002.c
#include "euler_solution.c"

int main() {
    execute(run);
}

long long fib(int n) {
    if (n == 0 || n == 1) {
        return n;
    }
    return fib(n - 1) + fib(n - 2);
}

long long calc() {
    long long sum = 0;
    int i = 0;
    while (1) {
        long curr = fib(i);
        if (curr >= 4000000) {
            break;
        }
        if (curr % 2 == 0) {
            sum += curr;
        }
        i++;
    }
    return sum;
}

long long fastCalc() {
    int x = 0;
    int y = 1;
    long long sum = 0;
    while (1) {
        int curr = x + y;
        x = y;
        y = curr;

        if (curr >= 4000000) {
            break;
        }

        if (curr % 2 == 0) {
            sum += curr;
        }
    }
    return sum;
}

long long run() {
    // ~80ms
    // return calc();

    // ~0ms
    return fastCalc();
}
